File: src/net/dinkla/raytracer/lights/PointLight.java
package net.dinkla.raytracer.lights;

import net.dinkla.raytracer.colors.Color;
import net.dinkla.raytracer.hits.Shade;
import net.dinkla.raytracer.math.Point3D;
import net.dinkla.raytracer.math.Ray;
import net.dinkla.raytracer.math.Vector3D;
import net.dinkla.raytracer.worlds.World;


public class PointLight extends Light {

    // emissive material
    public double ls;
    public Color color;
    public Point3D location;

    protected Color cachedL;

    public PointLight(Point3D location) {
        this.location = location;
        color = Color.WHITE;
        ls = 1.0;
        setShadows(true);
    }

    @Override
    public Color L(World world, Shade sr) {
        assert null != color;
        if (null == cachedL) {
            cachedL = color.mult(ls);
        }
        return cachedL;
    }

    @Override
    public Vector3D getDirection(Shade sr) {
        return new Vector3D(location.minus(new Vector3D(sr.getHitPoint()))).normalize();
    }

    @Override
    public boolean inShadow(World world, Ray ray, Shade sr) {
        double d = location.minus(ray.getO()).length();
        return world.inShadow(ray, sr, d);
    }

}

File: src/main/kotlin/net/dinkla/raytracer/lights/PointLight.kt
package net.dinkla.raytracer.lights

import net.dinkla.raytracer.colors.Color
import net.dinkla.raytracer.hits.Shade
import net.dinkla.raytracer.interfaces.Counter
import net.dinkla.raytracer.math.Point3D
import net.dinkla.raytracer.math.Ray
import net.dinkla.raytracer.math.Vector3D
import net.dinkla.raytracer.world.World

class PointLight(val location: Point3D = Point3D.ORIGIN,
                 val ls: Double = 1.0,
                 val color: Color = Color.WHITE,
                 override val shadows: Boolean = true) : Light {

    override fun L(world: World, sr: Shade): Color = color * ls

    override fun getDirection(sr: Shade): Vector3D = Vector3D(location - (Vector3D(sr.hitPoint))).normalize()

    override fun inShadow(world: World, ray: Ray, sr: Shade): Boolean {
        val d = (location - ray.origin).length()
        return world.inShadow(ray, sr, d)
    }
}
